File: python/knn.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 26 10:08:34 2025
Author: Chukwuka Chijioke Jerry

K-Nearest Neighbors (K-NN) Classifier on Social Network Ads
"""

# ==================== Import Libraries ====================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

# ==================== Import Dataset ====================
dataset = pd.read_csv('../data/social_network_ads.csv')
X = dataset.iloc[:, :2].values  # Age, EstimatedSalary
y = dataset.iloc[:, -1].values  # Purchased

# ==================== Train/Test Split ====================
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# ==================== Feature Scaling ====================
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# ==================== Fit the K-NN Model ====================
classifier = KNeighborsClassifier()
classifier.fit(X_train, y_train)

# ==================== Prediction and Evaluation ====================
y_pred = classifier.predict(X_test)

# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True, fmt='d', cbar=False)
plt.title('Confusion Matrix')
plt.tight_layout()
plt.show()

# Accuracy
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")

# ==================== Decision Boundary Visualization ====================

def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.arange(X_set[:, 0].min()-1, X_set[:, 0].max()+1, 0.01),
        np.arange(X_set[:, 1].min()-1, X_set[:, 1].max()+1, 0.01)
    )
    Z = classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape)

    plt.contourf(X1, X2, Z, alpha=0.3, cmap=ListedColormap(('red', 'green')))
    for i, j in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == j, 0], X_set[y_set == j, 1],
            color=['red', 'green'][i], label=j
        )
    plt.title(title)
    plt.xlabel('Age (scaled)')
    plt.ylabel('Estimated Salary (scaled)')
    plt.legend()
    plt.tight_layout()
    plt.show()

# ==================== Visualizing Training Set ====================
plot_decision_boundary(X_train, y_train, title='K-NN (Training Set)')

# ==================== Visualizing Test Set ====================
plot_decision_boundary(X_test, y_test, title='K-NN (Test Set)')

File: r/knn.R
# K-Nearest Neighbors (K-NN) for Social Network Ads

# ==================== Import Dataset ====================
dataset <- read.csv('../data/social_network_ads.csv')

# ==================== Split into Training and Test Sets ====================
library(caTools)
set.seed(123)  # for reproducibility
split <- sample.split(dataset$Purchased, SplitRatio = 0.8)
dataset_train <- subset(dataset, split == TRUE)
dataset_test <- subset(dataset, split == FALSE)

# ==================== Feature Scaling ====================
dataset_train[, -3] <- scale(dataset_train[, -3])
dataset_test[, -3] <- scale(dataset_test[, -3])

# ==================== Fit K-NN Model and Predict ====================
library(class)
y_pred <- knn(
  train = dataset_train[, -3],
  test = dataset_test[, -3],
  cl = dataset_train[, 3],
  k = 5
)

# ==================== Confusion Matrix ====================
cm <- table(dataset_test$Purchased, y_pred)
print(cm)

# ==================== Decision Boundary - Training Set ====================
set <- dataset_train
X1 <- seq(min(set$Age)-1, max(set$Age)+1, by = 0.01)
X2 <- seq(min(set$EstimatedSalary)-1, max(set$EstimatedSalary)+1, by = 0.01)
grid_set <- expand.grid(X1, X2)
colnames(grid_set) <- c('Age', 'EstimatedSalary')
y_grid <- knn(
  train = set[, -3],
  cl = set[, 3],
  k = 5,
  test = grid_set
)
par(mar = c(4.5, 4.5, 2.5, 1))  # slightly smaller margins
# Plot training set decision boundary
plot(
  set[, -3],
  main = 'K-NN (Training Set)',
  xlab = 'Age',
  ylab = 'Estimated Salary',
  xlim = range(X1),
  ylim = range(X2)
)
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch='.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set[, -3], pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

# ==================== Decision Boundary - Test Set ====================
set <- dataset_test
X1 <- seq(min(set$Age)-1, max(set$Age)+1, by = 0.01)
X2 <- seq(min(set$EstimatedSalary)-1, max(set$EstimatedSalary)+1, by = 0.01)
grid_set <- expand.grid(X1, X2)
colnames(grid_set) <- c('Age', 'EstimatedSalary')
y_grid <- knn(
  train = set[, -3],
  cl = set[, 3],
  k = 5,
  test = grid_set
)

# Plot test set decision boundary
plot(
  set[, -3],
  main = 'K-NN (Test Set)',
  xlab = 'Age',
  ylab = 'Estimated Salary',
  xlim = range(X1),
  ylim = range(X2)
)
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch='.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(
  set[, -3], pch = 21,
  bg = ifelse(set[, 3] == 1, 'green4', 'red3')
)
